### ISSUE TITLE
TST Improve error message of assert_produces_warning

### ISSUE THREAD

**Issue by MarcoGorelli** [pandas-dev/pandas#42103]
Currently, when using `assert_produces_warning` with `match=`, the error message is not helpful. e.g.:
```
E           AssertionError: Did not see warning 'FutureWarning' matching The default value of regex wilt change from True to False in a future version\.$
```

I think it should show the diff of the message, like pytest does

**Comment by calvinsomething** [comment]
take

**Comment by MarcoGorelli** [comment]
closed by #42181
